### ISSUE TITLE
Suite of integration tests for Minion

### ISSUE THREAD

**Issue by gskorokhod** [conjure-cp/conjure-oxide#249]
We need a large suite of problems to solve with minion and compare our solutions to conjure (see #247).

Having a large test suite would probably flag up some issues (unsupported Essence features, unsupported Minion constraints, issues with the rewriting engine, issues in minion_rs) and hopefully provide more direction for further work

**Comment by ChrisJefferson** [comment]
This is probably a long term project, but I strongly believe fuzzing is the only way to get top quality coverage. Once we had a program which could generate valid minion inputs, and tested with those, we found dozens of bugs in minion I'd never have found otherwise.

**Comment by ozgurakgun** [comment]
I agree that fuzzing will be good and I agree that we are not there yet :)

We probably want a native parser first, before we get to fuzzing.

**Comment by niklasdewally** [comment]
> I agree that fuzzing will be good and I agree that we are not there yet :)
> 
> We probably want a native parser first, before we get to fuzzing.

We can fuzz into rust types fairly easily using the abitrary crate!

**Comment by niklasdewally** [comment]
See also 
https://github.com/conjure-cp/conjure-oxide/issues/231

**Comment by niklasdewally** [comment]
Not sure what level we are talking at here, but my plan for minion_rs was to use the standard integration tests for "valid models" and "right solutions" and then test safety and ffi and the c++ with fuzzing. If we fuzz the bindings well, we may not need to fuzz in conjure-oxide itself for minion, as minion_rs would be more or less safe (in the memory and ub sense)

**Comment by gskorokhod** [comment]
> I agree that fuzzing will be good and I agree that we are not there yet :)
> 
> 
> 
> We probably want a native parser first, before we get to fuzzing.

I think long term it would be useful to have both a Rust native parser for Essence and pretty printing our AST back to essence. Then it would probably be relatively easy to procedurally generate some valid models and pretty print them to an Essence file that we can feed to old conjure to compare results?

**Comment by ozgurakgun** [comment]
this is ongoing work that probably concerns a few people. at least @niklasdewally and @EEJDempster

**Comment by niklasdewally** [comment]
I am not sure what exactly the scope of this issue is.

I am adding essence prime tests and tests targeting specific Minion
constraints as part of my standard process for adding rules and doesn't
need to be done separately.

The only other specific todo I can think of for Minion correctness is
fuzzing minion_rs / Conjure Oxide and occasionally checking coverage to
see what test cases I've missed.

The former should be in seperate issues..

**Comment by niklasdewally** [comment]
Related: #171 tracks low level `eprime` integration tests that target specific minion constraints.

**Comment by ozgurakgun** [comment]
@niklasdewally is on top of this. And performance comparison is something @EEJDempster will look into. The issue as it's written is too general to be useful, closing.
